salt/config/__init__.py

```diff
@@ -3325,13 +3369,19 @@ def get_id(opts, cache_minion_id=False):
         log.debug('Guessing ID. The id can be explicitly set in {0}'
                   .format(os.path.join(salt.syspaths.CONFIG_DIR, 'minion')))
 
-    newid = salt.utils.network.generate_minion_id()
+    if opts.get('minion_id_type', 'str') == 'func':
+        newid = eval_minion_id_func(opts)
+    else:
+        newid = salt.utils.network.generate_minion_id()
```

Comment:
I'd not design it this way. The eval_minion_id_func should just fall-back inside. So you won't need to check this numerous times here/there, but just do it once and for all:
```python
newid = eval_minion_id_func(opts)
```
And that will do `generate_minion_id()` inside in case opts aren't `func`. But again, I would not use this `func`/`type` thing, but rather just a module name directly. So my code would be like:

```python
custom_module = opts.get('minion_id_function')
newid = minion_id_mod[custom_module](*args, **kwargs) if custom_module \
            else salt.utils.network.generate_minion_id()
```